### PR TITLE
ENH: Allow future chains to only use certain delivery months.

### PR DESCRIPTION
**Pull request by ehebert**
To support contracts such as `PL` which should roll from F->J->N->V, add the
ability to pass a predicate function to the ordered contract chain contstrution
which returns `True` if the contract is allowed in the chain.

### CODE REVIEW

**Comment by ehebert** [comment]
A graph of continuous `PL` volume without this patch applied.

pl-without-chain-predicate

This is with this patch applied. 

pl-with-chain-predicate

File: zipline/assets/continuous_futures.pyx [hunk 1]
```diff
@@ -295,12 +312,13 @@ cdef class OrderedContracts(object):
...
     cdef readonly object root_symbol
-    cdef readonly object head_contract
+    cdef readonly object _head_contract
     cdef readonly dict sid_to_contract
     cdef readonly int64_t _start_date
     cdef readonly int64_t _end_date
+    cdef readonly object chain_predicate
 
-    def __init__(self, object root_symbol, object contracts):
+    def __init__(self, object root_symbol, object contracts, object chain_predicate=None):
```
Comment (dmichalowicz) [inline zipline/assets/continuous_futures.pyx]: Same here with the docstring

File: zipline/assets/assets.py [hunk 2]
```diff
@@ -193,7 +197,7 @@ class AssetFinder(object):
     PERSISTENT_TOKEN = "<AssetFinder>"
 
     @preprocess(engine=coerce_string_to_eng)
-    def __init__(self, engine):
+    def __init__(self, engine, future_chain_predicates=CHAIN_PREDICATES):
```
Comment (dmichalowicz) [inline zipline/assets/assets.py]: Thinking more, this feels like it doesn't quite belong as a parameter to `AssetFinder`, but I'm not sure how else to refactor it. Maybe make this an optional parameter to `get_ordered_contracts`?

File: zipline/assets/continuous_futures.pyx [hunk 3]
```diff
@@ -37,6 +39,21 @@ import warnings
 from zipline.utils.calendars import get_calendar
 
 
+def delivery_predicate(codes, contract):
+    # This relies on symbols that are construct following a pattern of
+    # root symbol + delivery code + year, e.g. PLF16
+    # This check would be more robust if the future contract class had
+    # a 'delivery_month' member.
+    delivery_code = contract.symbol[2]
```
Comment (dmichalowicz) [inline zipline/assets/continuous_futures.pyx]: Is the `delivery_code` guaranteed to be the third character of the symbol, i.e. the root symbol is always 2 characters?

File: tests/test_continuous_futures.py [hunk 4]
```diff
@@ -144,7 +152,34 @@ def make_futures_info(self):
...
+                         Timestamp('2016-10-19', tz='UTC')],
+            'notice_date': [Timestamp('2016-01-11', tz='UTC'),
+                            Timestamp('2016-02-13', tz='UTC'),
+                            Timestamp('2016-03-10', tz='UTC')],
+            'expiration_date': [Timestamp('2016-01-11', tz='UTC'),
+                                Timestamp('2016-02-13', tz='UTC'),
+                                Timestamp('2016-03-10', tz='UTC')],
+            'auto_close_date': [Timestamp('2016-01-11', tz='UTC'),
+                                Timestamp('2016-02-13', tz='UTC'),
+                                Timestamp('2016-03-10', tz='UTC')],
```
Comment (dmichalowicz) [inline tests/test_continuous_futures.py]: Is there a particular method to contrive these dates, or are they arbitrary?

**Comment by dmichalowicz** [comment]
@ehebert Done with a first round of review, mostly had a few questions but overall looks good. Also thanks for the screenshots

File: zipline/assets/assets.py (hunk 2 above)
Comment (ehebert) [inline zipline/assets/assets.py]: Yes.

**Pull request merged**